### PR TITLE
Telegram bot operator

### PR DESCRIPTION
**Pull request by byymster**
## Description

Add PyCryptoBot K8S operator
Add ability for telegram_bot to run pycryptobot is separate K8S Jobs 

Video example:

[Alt text](https://www.youtube.com/watch?v=qOJ8MXU7Bo8)

## Type of change

Please make your selection.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change requires a new dependency
- [ ] Code Maintainability / comments
- [ ] Code Refactoring / future-proofing

## How Has This Been Tested?

I tested it locally in k3d leave for a night with live:0, no issues detected

## Checklist:

- [X] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added pytest unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: docs/RunningTelegramBotInK8S.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+resources:
+ requests:
+   cpu: 500m
+   memory: 1Gi
+ limits:
+   cpu: 2
+   memory: 2.5Gi
+```
+5. It's highly recommend to set `"disablecleanfilesonexit": 1`, in other case if pod gracefully restart you'll lose running crypto-bots
+
```
Comment (loomsen) [inline docs/RunningTelegramBotInK8S.md:30]: maybe we should provide a complete example or people could ask if these should go into separate yamls?

**Comment by loomsen** [comment]
Cool stuff, I like it :) Haven't had the change to test, but great idea. Thank you!

**Comment by byymster** [comment]
> Cool stuff, I like it :) Haven't had the change to test, but great idea. Thank you!

Thanks, it's even can executed on few RPi's, if make k8s cluster on them
I'm using k3s for that, it's lightweight and fast

File: docs/RunningTelegramBotInK8S.md (hunk 1 above)
Comment (byymster) [reply docs/RunningTelegramBotInK8S.md:30]: now is more clear?

Comment (loomsen) [reply docs/RunningTelegramBotInK8S.md:30]: Oh what I meant is providing a full working `values.yaml`. 

like:

Fully working example helm values file:
```
telegramBot:
  persistence:
    enabled: true
    storageClass: somestorageclass
command:
  - python3
  - -u
  - telegram_bot.py
  - --restart-on-init
resources:
 requests:
   cpu: 500m
   memory: 1Gi
 limits:
   cpu: 2
   memory: 2.5Gi
```

Something like that maybe?

Comment (loomsen) [reply docs/RunningTelegramBotInK8S.md:30]: Although, when I think about it, we can expect people to know how to use helm and kubernetes when they want to go that way I guess.

Comment (byymster) [reply docs/RunningTelegramBotInK8S.md:88]: Like this?

Comment (loomsen) [reply docs/RunningTelegramBotInK8S.md:88]: Yep, thanks, I like it and I think it's more clear. :+1:

**Review by loomsen — APPROVED** [review]

**Comment by whittlem** [comment]
@byymster and @markhollingworth-worthit, thanks for your PRs. Can you please look at the conflicts in telegram_bot.py?

**Comment by whittlem** [comment]
What does "chart" and "pycryptobot-operator" actually do?

**Comment by byymster** [comment]
> What does "chart" and "pycryptobot-operator" actually do?

This 2 links can explain better than me: [helm charts](https://www.bmc.com/blogs/kubernetes-helm-charts/), [Kubernetes operator](https://www.redhat.com/en/topics/containers/what-is-a-kubernetes-operator)

in general I changed chart a little to be able deploy configuration that can run telegram_bot
Made operator and custom CRD to be able run pycryptobot as custom Kubernetes API entity
Operator based on list of the entities(pycryptobots) deploys helm chart with specific configuration for each entities
Rewrite a little telegram_bot to be able create/delete  pycryptobot CRD instead of running subprocess
For K8S it's native way how to make dynamic distributed applications

**Comment by byymster** [comment]
@whittlem I need some time for tests because after reverting merged code it lost my previous changes.
I'll review everything and retest it and ping you after.

**Comment by loomsen** [comment]
@whittlem These changes are related to running the bot on kubernetes. Nothing else should be affected. (remember you wanted to create an article about how to run it in kubernetes when I initially added the helm chart? :) )

**Comment by byymster** [comment]
@whittlem looks ok, I run few tests and it works. I'll leave it for a night will see how it's going

**Comment by notoriousBLP** [comment]
been wanting this since i showed up a month ago...thanks @byymster

**Pull request merged**